Mathlib/Analysis/Convex/Radon.lean

```diff
@@ -49,3 +51,54 @@ theorem radon_partition (h : ¬ AffineIndependent 𝕜 f) :
     (fun _i hi ↦ Set.mem_image_of_mem _ fun hi' ↦ ?_)
   · linarith only [hI, hJI]
   · exact (mem_filter.mp hi').2.not_lt (mem_filter.mp hi).2
+
+
+open FiniteDimensional
+
+variable [FiniteDimensional 𝕜 E]
+
+/-- **Helly's theorem on convex sets**: If `F` is a finite family of convex sets in a vector space
+of finite dimension `d`, and any `d + 1` sets of `F` intersect, then all sets of `F` intersect. -/
+theorem helly_theorem (F : Set (Set E)) {hF_fin : Set.Finite F}
+    (h_card : (finrank 𝕜 E) + 1 ≤ ncard F) (h_convex : ∀ X ∈ F, Convex 𝕜 X)
+    (h_inter : ∀ G : Set (Set E), (G ⊆ F) → (ncard G = (finrank 𝕜 E) + 1) →
+    (⋂₀ G).Nonempty) : (⋂₀ F).Nonempty := by
+  obtain ⟨n, hn⟩ : ∃ n : ℕ, ncard F = n := ⟨ncard F, rfl⟩ -- for induction on ncard F
+  rw [hn] at h_card
+  induction' n, h_card using Nat.le_induction with k h_card hk generalizing F
+  exact h_inter F (Eq.subset rfl) hn
+  have h1 : ∀ X ∈ F, (⋂₀ (F \ {X})).Nonempty := by
+    intro X hX
+    apply @hk _ (Finite.diff hF_fin {X})
+    · exact fun Y hY ↦ h_convex Y (mem_of_mem_diff hY)
+    · exact fun G hG_1 hG_2 ↦ h_inter G (Subset.trans hG_1 (diff_subset F {X})) hG_2
+    · rw [ncard_diff_singleton_of_mem hX hF_fin, Nat.sub_eq_of_eq_add hn]
+  let a : F → E := fun X ↦ (h1 X (Subtype.mem X)).some
+  have h2 : ¬AffineIndependent 𝕜 a := by
+    have : Fintype ↑F := Finite.fintype hF_fin -- for instance inferring
+    rw [←finrank_vectorSpan_le_iff_not_affineIndependent 𝕜 a (n := (k - 1))]
+    · exact (Submodule.finrank_le (vectorSpan 𝕜 (Set.range a))).trans (Nat.le_pred_of_lt h_card)
+    · rw [←Finite.card_toFinset hF_fin, ←ncard_eq_toFinset_card F hF_fin, hn,
+      ←Nat.sub_add_comm (Nat.one_le_of_lt h_card)]
+      rfl
+  obtain ⟨I, p, h4_I, h4_Ic⟩ := Convex.radon_partition h2
+  use p
+  rw [mem_sInter]
+  intro X hX
+  have h3 (X Y : Set E) (hX : X ∈ F) (hY : Y ∈ F) (hneq : X ≠ Y) : a ⟨Y, hY⟩ ∈ X := by
+    apply @mem_of_subset_of_mem _ (⋂₀ (F \ {Y})) _
+    apply sInter_subset_of_mem
+    · simp only [mem_diff, hX, mem_singleton_iff, hneq, not_false_eq_true, and_self]
+    · exact (h1 Y hY).some_mem
```

Comment:
with the `lift` suggested above, this can be golfed quite a bit, and to be honest, `mem_of_subset_of_mem` is kind of a superfluous lemma. You can just replace it with a proof of the subset fact, and then apply it to a proof of membership in the smaller set.